rust_src/src/textprop.rs

```diff
@@ -30,7 +30,42 @@ pub fn get_char_property(position: LispNumber, prop: LispObject, object: LispObj
 /// If POSITION is at the end of OBJECT, the value is nil.
 #[lisp_fn(min = "2")]
 pub fn get_text_property(position: LispNumber, prop: LispObject, object: LispObject) -> LispObject {
-    unsafe { textget(Ftext_properties_at(position.into(), object), prop) }
+    unsafe { textget(text_properties_at(position.into(), object), prop) }
+}
+
+/// Return the list of properties of the character at POSITION in OBJECT.
+/// If the optional second argument OBJECT is a buffer (or nil, which means
+/// the current buffer), POSITION is a buffer position (integer or marker).
+/// If OBJECT is a string, POSITION is a 0-based index into it.
+/// If POSITION is at the end of OBJECT, the value is nil.
+#[lisp_fn(min = "1")]
+pub fn text_properties_at(mut position: LispObject, mut object: LispObject) -> LispObject {
+    if object.is_nil() {
+        object = ThreadState::current_buffer().into();
+    }
+
+    let i: LispIntervalRef = LispIntervalRef::new(unsafe {
```

Comment:
`validate_interval_range` is used as a constructor in the C code except in one place where it is only used for the validation. A more appropriate design for `LispIntervalRef` I think would be to add a `new_from_position()` factory method which takes the object and the position (only once). In this use case the final option is always false in the C code.

`let i: LispIntervalRef::new_from_position(object, position);`

This does however lose the "validated" which is likely important for the reader. Should the name be `new_validated_from_position`? `new_validated_position`? `new_position_with_validation`?

Reply:
I have it as `from_valid_range` locally, with the validation itself separated into another function that's called internally. It seems to work in that the tests still pass, but I'm not 100% clear on what the XSETFASTINT macro does on the C version. It's being called to assign the value of `*begin` to itself, but I couldn't quite parse the bitwise operations it does. Is that where they increment these numbers by 1?

Reply:
For Rust we can essentially read `XSETFASTINT` as `let x: LispObject = LispObject::from(some_number);`.